### PR TITLE
Add P2S configuration with AAD enabled

### PR DESCRIPTION
**Pull request by oleksiimorozenko**
It’s not possible to set up P2S connections using terraform, it can be only made manually .
If VPN GW was created using this module, and manual changes have been made to its configuration, it will try to sync infrastructure with the state, i.e. wipe all the manual changes out like this:
```
      - custom_route {
          - address_prefixes = [
              - "10.0.0.0/16",
            ] -> null
        }


      - vpn_client_configuration {
          - aad_audience         = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -> null
          - aad_issuer           = "https://sts.windows.net/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/" -> null
          - aad_tenant           = "https://login.microsoftonline.com/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/" -> null
          - address_space        = [
              - "172.16.0.0/24",
            ] -> null
          - vpn_auth_types       = [
              - "AAD",
            ] -> null
          - vpn_client_protocols = [
              - "OpenVPN",
            ] -> null
        }
``` 
This PR's changes can prevent it and P2S connection can be made entirely in terraform.
Also, there’s a change advertising’s route to defined networks which should be accessed through the VPN service.

P.S. Since AAD authorization is available with OpenVPN only, those two settings are hardcoded.

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -137,3 +137,26 @@ EOD
...
+```
+map(
+  aad_audience          string                # The client id of the Azure VPN application
+  aad_issuer            string                # The STS url for your tenant
+  aad_tenant            string                # AzureAD Tenant URL
+  address_space         list(string)          # The address space out of which IP addresses for vpn clients will be taken
+)
+```
+EOD
+  type = map(any)
```
Comment (Shr3ps) [inline variables.tf]: Can you change this to use a typed object?

File: variables.tf [hunk 2]
```diff
@@ -137,3 +137,26 @@ EOD
...
+  default = {}
```
Comment (Shr3ps) [inline variables.tf]: What about null by default?

File: variables.tf [hunk 3]
```diff
@@ -137,3 +137,26 @@ EOD
...
+}
+
+variable "additional_routes_to_advertise" {
+  description = "Additional routes to advertise"
+  default = []
```
Comment (Shr3ps) [inline variables.tf]: You should trigger `terraform fmt` ;)

**Review by Shr3ps — CHANGES_REQUESTED** [review]
Thanks for your contribution!

File: variables.tf (hunk 1 above)
Comment (oleksiimorozenko) [reply variables.tf]: Done

File: variables.tf (hunk 2 above)
Comment (oleksiimorozenko) [reply variables.tf]: Changed to `null`

File: variables.tf (hunk 3 above)
Comment (oleksiimorozenko) [reply variables.tf]: Done

File: r-vpn.tf [hunk 4]
```diff
@@ -37,6 +37,27 @@ resource "azurerm_virtual_network_gateway" "public_virtual_network_gateway" {
...
+    for_each = var.vpn_aad_client_configuration != null ? [var.vpn_aad_client_configuration] : []
+    iterator = vpn
+
+    content {
+      aad_audience         = vpn.value.aad_audience
+      aad_issuer           = vpn.value.aad_issuer
+      aad_tenant           = vpn.value.aad_tenant
+      address_space        = [vpn.value.address_space]
+      vpn_auth_types       = ["AAD"]
+      vpn_client_protocols = ["OpenVPN"]
```
Comment (jmapro) [inline r-vpn.tf]: With hardcoded value, we can't use any other protocol.
If `AAD` is not compatible with anything else than `OpenVPN` you can use variable validation or `pre_condition` checks.

https://www.terraform.io/language/expressions/custom-conditions

File: variables.tf [hunk 5]
```diff
@@ -137,3 +137,32 @@ EOD
...
+  address_space         list(string)          # The address space out of which IP addresses for vpn clients will be taken
+)
+Additional information can be found at (https://learn.microsoft.com/en-us/azure/vpn-gateway/openvpn-azure-ad-tenant#enable-authentication)
+```
+EOD
+  type = object({
+    aad_audience  = string                 # The client id of the Azure VPN application
+    aad_issuer    = string                 # The STS url for your tenant
+    aad_tenant    = string                 # AzureAD Tenant URL
+    address_space = optional(list(string)) # The address space out of which IP addresses for vpn clients will be taken
```
Comment (jmapro) [inline variables.tf:168]: Please add all other possible configuration options.

File: variables.tf [hunk 6]
```diff
@@ -137,3 +137,32 @@ EOD
...
+  aad_tenant            string                # AzureAD Tenant URL
+  address_space         list(string)          # The address space out of which IP addresses for vpn clients will be taken
+)
+Additional information can be found at (https://learn.microsoft.com/en-us/azure/vpn-gateway/openvpn-azure-ad-tenant#enable-authentication)
+```
+EOD
+  type = object({
+    aad_audience  = string                 # The client id of the Azure VPN application
+    aad_issuer    = string                 # The STS url for your tenant
+    aad_tenant    = string                 # AzureAD Tenant URL
```
Comment (jmapro) [inline variables.tf]: Should be optional since we don't want to force OpenVPN as the only protocol.

**Review by jmapro — CHANGES_REQUESTED** [review]

File: r-vpn.tf [hunk 7]
```diff
@@ -37,6 +37,27 @@ resource "azurerm_virtual_network_gateway" "public_virtual_network_gateway" {
     }
   }
 
+  dynamic "custom_route" {
+    for_each = var.additional_routes_to_advertise
```
Comment (Shr3ps) [inline r-vpn.tf]:
```suggestion
    for_each = var.additional_routes_to_advertise != null ? [1] : []
```

**Review by Shr3ps — CHANGES_REQUESTED** [review]

**Review by Shr3ps — APPROVED** [review]

**Comment by Shr3ps** [comment]
Released in latest `v7.1.0` :)

**Pull request merged**